File: java-rent/src/main/java/de/richargh/billiondollar/Inventory.java
package de.richargh.billiondollar;

import java.util.Map;
import java.util.Optional;
import java.util.stream.*;

public class Inventory {

    private final Map<ItemId, Item> items;

    public Inventory(Item... items){
        this.items = Stream.of(items).collect(Collectors.toMap(Item::id, item -> item));
    }

    public Optional<Item> findById(ItemId id){
        return Optional.ofNullable(items.get(id));
    }

    public boolean rent(Item item, RenterId renterId){
        Item rentedItem = new Item(item.id(), item.name(), Optional.of(renterId));
        items.put(rentedItem.id(), rentedItem);
        return true;
    }

}

File: python-rent/src/renting/Inventory.py
from src.renting.items import Item, AvailableItem, RentedItem
from src.renting.ItemId import ItemId
from src.renting.RenterId import RenterId


class Inventory:
    _items: dict[ItemId, Item] = {}

    def __init__(self, *items: Item):
        self._items = {item.id: item for item in items}

    def find_by_id(self, item_id: ItemId) -> Item | None:
        return self._items.get(item_id, None)

    def rent(self, item: AvailableItem, renter_id: RenterId) -> None:
        rented_item = RentedItem(item.id, item.name, renter_id)
        self._items[rented_item.id] = rented_item
